### PR TITLE
Darkside treestate support

### PR DESCRIPTION
**Pull request by pacu**
Closes #389 

How to test this:

compile the code with `make`

then start darksidewalletd
`./lightwalletd --log-file /dev/stdout --darkside-very-insecure  --darkside-timeout 1000 --gen-cert-very-insecure --data-dir . --no-tls-very-insecure`

reset the chain
`grpcurl -plaintext -d '{ "saplingActivation" : 663150, "branchID" : "2bb40e60", "chainName" : "main"}' localhost:9067 cash.z.wallet.sdk.rpc.DarksideStreamer/Reset`

stage the tree state

````
grpcurl -plaintext -d '{ 
  "network": "main",
  "height": "663150",
  "hash": "0000000002fd3be4c24c437bd22620901617125ec2a3a6c902ec9a6c06f734fc",
  "time": 1576821833,
  "saplingTree": "01ec6278a1bed9e1b080fd60ef50eb17411645e3746ff129283712bc4757ecc833001001b4e1d4a26ac4a2810b57a14f4ffb69395f55dde5674ecd2462af96f9126e054701a36afb68534f640938bdffd80dfcb3f4d5e232488abbf67d049b33a761e7ed6901a16e35205fb7fe626a9b13fc43e1d2b98a9c241f99f93d5e93a735454073025401f5b9bcbf3d0e3c83f95ee79299e8aeadf30af07717bda15ffb7a3d00243b58570001fa6d4c2390e205f81d86b85ace0b48f3ce0afb78eeef3e14c70bcfd7c5f0191c0000011bc9521263584de20822f9483e7edb5af54150c4823c775b2efc6a1eded9625501a6030f8d4b588681eddb66cad63f09c5c7519db49500fc56ebd481ce5e903c22000163f4eec5a2fe00a5f45e71e1542ff01e937d2210c99f03addcce5314a5278b2d0163ab01f46a3bb6ea46f5a19d5bdd59eb3f81e19cfa6d10ab0fd5566c7a16992601fa6980c053d84f809b6abcf35690f03a11f87b28e3240828e32e3f57af41e54e01319312241b0031e3a255b0d708750b4cb3f3fe79e3503fe488cc8db1dd00753801754bb593ea42d231a7ddf367640f09bbf59dc00f2c1d2003cc340e0c016b5b13"
}' localhost:9067 cash.z.wallet.sdk.rpc.DarksideStreamer/AddTreeState
````

obtaining the TreeState

`grpcurl -plaintext -d '{ "height" : 663150 }' localhost:9067 cash.z.wallet.sdk.rpc.CompactTxStreamer/GetTreeState`

removing a tree state
`grpcurl -d '{ "height" : 663150 }' localhost:9067 cash.z.wallet.sdk.rpc.DarksideStreamer/RemoveTreeState`

clearing all tree states from cache

`grpcurl -d '{}' localhost:9067 cash.z.wallet.sdk.rpc.DarksideStreamer/ClearAllTreeStates`

# known issues
somehow the get tree state always attempts to call Zcashd
`grpcurl -plaintext -d '{ "height" : 663150 }' localhost:9067 cash.z.wallet.sdk.rpc.CompactTxStreamer/GetTreeState`

```
pacu@pacu lightwalletd % grpcurl -plaintext -d '{ "height" : 663150 }' localhost:9067 cash.z.wallet.sdk.rpc.CompactTxStreamer/GetTreeState
ERROR:
  Code: Unknown
  Message: zcashd did not return treestate
```

### CODE REVIEW

**Comment by LarryRuane** [comment]
Yes, this definitely should be fixed, and it's not hard to do. 

One suggestion I have, would it be okay if `cash.z.wallet.sdk.rpc.DarksideStreamer/Reset` clears the treestate, instead of having a separate `ClearAllTreeStates` gRPC? That gRPC currently clears all existing darkside state, so it would be simpler if that's how we clear all of the treestates.

**Comment by LarryRuane** [comment]
>I was thinking of TreeStates being reused across tests ...

Ah, I see; also I just noticed that you're already implemented this in this PR, so I'm totally fine with that. All this stuff is to make you're (wallet devs) lives' easier, so however you think it should work is what we should do!

I'll try to get to this (review what's here, rebase, make any additional tweaks) soon.

**Comment by pacu** [comment]
@LarryRuane I'm working on this again and I see that I'm not overriding the getTreeState method at the right level. 

I'm returning the Tree state here

````
case "z_gettreestate":
		var heightStr string
		err := json.Unmarshal(params[0], &heightStr)
		if err != nil {
			return nil, errors.New("failed to parse gettreestate request")
		}

		height, err := strconv.Atoi(heightStr)
		if err != nil {
			return nil, errors.New("error parsing height as integer")
		}

		treeState := state.stagedTreeStates[uint64(height)]

		if treeState == nil {
			return nil, errors.New(fmt.Sprint(
				"there is no TreeState Staged for height \"",
				height,
				"\". Stage it using AddTreeState() first"))
		}

		return json.Marshal(treeState)

````

but this actually makes the real lwd be executed instead and the expected tree state format is the one returned by zcashd

````
ZcashdRpcReplyGettreestate struct {
		Height  int
		Hash    string
		Time    uint32
		Sapling struct {
			Commitments struct {
				FinalState string
			}
			SkipHash string
		}
		Orchard struct {
			Commitments struct {
				FinalState string
			}
			SkipHash string
		}
	}
````

shall I be attempting to override the compactStreamer `gettreestate` call instead by adding a

````
func darksideRawRequest(method string, params []json.RawMessage) (json.RawMessage, error) {
	switch method {
	/// other cases
        /// ...
	case "gettreestate":
		return nil, errors.New("Hello this is working")
````

instead?

**Comment by pacu** [comment]
@LarryRuane I had a super old lightwalletd repo! it dated when this repo was over on zcash-hackworks. I fixed that. 

I consider it ready for review.

**Review by LarryRuane — COMMENTED** [review]
I tested it, looks good, other than a small suggestion. 

One other small thing I noticed is that if you `AddTreeState` with the `network` field set to `"main"` (as in this PR's description), it fetches back as `"darkside"`. (That's because the `zcashd` RPC doesn't return that field, and it's [added](https://github.com/zcash/lightwalletd/blob/bbe8d14ecc8d73beffc54193c716f146f658f243/frontend/service.go#L230) by the `GetTreeState` gRPC handler.) It's not really a problem; it's not necessary to even specify `"network"` when you add a tree state. Maybe no change is needed for this.

File: common/darkside.go [hunk 1]
```diff
@@ -684,3 +730,18 @@ func DarksideClearAddressUtxos() error {
...
 	return nil
 }
+
+func DarksideClearAllTreeStates() error {
+	state.stagedTreeStates = make(map[uint64]*DarksideTreeState)
+	return nil
+}
+
+func DarksideAddTreeState(arg DarksideTreeState) error {
+	state.stagedTreeStates[arg.Height] = &arg
```
[the remark this answers is not in the document]
Comment (pacu) [reply common/darkside.go:740]: I think returning an educative error is best! I'll do that! Thanks Larry!

**Comment by LarryRuane** [comment]
This looks good, Pacu, but 2 things, can you squash the latest commit, and also I just noticed that the commit message for the first commit seems to be wrong, can you check?

**Comment by pacu** [comment]
> This looks good, Pacu, but 2 things, can you squash the latest commit, and also I just noticed that the commit message for the first commit seems to be wrong, can you check?

Done! thanks for catching that! My repo was quite broken and a rebase had mixed a bunch of stuff.

**Pull request merged**